subprojects/resources-http/src/main/java/org/gradle/internal/resource/transport/http/HttpClientHelper.java

```diff
@@ -128,25 +131,81 @@ private HttpClientResponse performHttpRequest(HttpRequestBase request, HttpConte
         httpContext.removeAttribute(REDIRECT_LOCATIONS);
         LOGGER.debug("Performing HTTP {}: {}", request.getMethod(), request.getURI());
+        validateUrl(request.getURI());
         try {
             CloseableHttpResponse response = getClient().execute(request, httpContext);
             return toHttpClientResponse(request, httpContext, response);
         } catch (IOException e) {
+            validateRedirectChain(httpContext);
             URI lastRedirectLocation = getLastRedirectLocation(httpContext);
             throw (lastRedirectLocation == null) ? e : new FailureFromRedirectLocation(lastRedirectLocation, e);
         }
     }
 
     private HttpClientResponse toHttpClientResponse(HttpRequestBase request, HttpContext httpContext, CloseableHttpResponse response) {
+        validateRedirectChain(httpContext);
         URI lastRedirectLocation = getLastRedirectLocation(httpContext);
         URI effectiveUri = lastRedirectLocation == null ? request.getURI() : lastRedirectLocation;
         return new HttpClientResponse(request.getMethod(), effectiveUri, response);
     }
 
-    @SuppressWarnings("unchecked")
-    private URI getLastRedirectLocation(HttpContext httpContext) {
-        List<URI> redirectLocations = (List<URI>) httpContext.getAttribute(REDIRECT_LOCATIONS);
-        return (redirectLocations == null || redirectLocations.isEmpty()) ? null : Iterables.getLast(redirectLocations);
+    private void validateUrl(URI url) {
+        /*
+         * For security purposes this intentionally requires a user to opt-in to http on case by case basis.
+         * We do not want to have a system/gradle property that allows a universal disable of this check.
+         */
+        if (settings.allowInsecureProtocol()) {
+            return;
+        }
+        /*
+         * TL;DR: http://127.0.0.1 will bypass this validation, http://localhost will fail this validation.
+         *
+         * Hundreds of Gradle's integration tests use a local web-server to test logic that relies upon
+         * this HttpClientHelper.
+         * Changing all of those tests so that they use a KeyStore would have been impractical.
+         * Instead, the test fixture was updated to use 127.0.0.1 when making HTTP requests.
+         *
+         * This allows tests that still want to exercise the deprecation logic to use http://localhost
+         * which will bypass this check and trigger the validation.
+         *
+         * It's important to note that the only way to create a certificate for an IP address is to bind
+         * the IP address as a 'Subject Alternative Name' which was deemed far too complicated for our test
+         * use case.
+         *
+         * Additionally, in the rare case that a user or a plugin author truly needs to test with a localhost
+         * server, they can use http://127.0.0.1
+         */
+        if ("127.0.0.1".equals(url.getHost())) {
```

Comment:
what about IPV6?

Reply:
This bypass is pretty much for our testing exclusively. It's not even going to be an externally documented feature.

If it's important, I can update this if you think I should.

Reply:
Not important for me, just wanted to point out that it's a possibility, too.